### DISCUSSION TITLE
Building as a wasm target

### DISCUSSION THREAD

**Discussion by prasunsrivastava** [gorules/zen#239] [Q&A]
Is it possible to build the library as a wasm binary? If yes, how do we go about doing that? Can anyone guide me on how to do this?

Looking forward to replies.

Thanks,
Prasun.

**Comment by ivanmiletic** [comment]
Hi @prasunsrivastava at the moment ZenEngine cant compile to WASM due to our dependency with QuickJS used for Function Nodes. Our Zen Expression language can compile and we have it in a production with one of our Enterprise customers.

We have a plan to cover other platforms and architectures such as iOS, Android and WASM but it is not planned for 2024.

**Reply by stefan-gorules** [reply]
Hi @prasuncurio

Our enterprise customers use ZEN expressions inside WASM (https://www.npmjs.com/package/@gorules/zen-engine-wasm) for form validation and for running custom rules business suites (without functions) or conditional content delivery.

You can define custom business rules suites using a custom format e.g.

```json5
{
  "callOuts": [
    {
      "title": "Grab seat next to window",
      "description": "Enjoy beautiful view",
      "showIf": "len(seat) == 0 and seat[0].type == 'adult'"
    },
    {
      "title": "Travel together",
      "description": "Book seats together with your family",
      "showIf": "some(seat, #.type == 'child')"
    },
  ]
}
``` 

This way you can have custom format generated by some UI, and available across both frontend and backend for smooth responsivness.

**Reply by prasunsrivastava** [reply]
@stefan-gorules This is pretty interesting. So, you mean to say the expression language can be used to create a custom rules engine that can allow us to create forms for taking inputs and render conditional content based on those responses? Currently we are using zen engine along with another in-house form-engine to drive this sort of behavior. Basically, we take inputs using the forms rendered through that form-engine and then, run the rules generated from the zen engine to get the recommendations. We never thought of using the expression language to do that.

**Reply by stefan-gorules** [reply]
Yep, you can absolutely do that as well. For example your form schema might look like:

```json5
{
  version: '1.0.0',
  elements: [
    {
      name: 'secretCode',
      label: 'Secret code',
      type: 'secret',
      required: true
    },
    {
      name: 'confirmSecretCode',
      label: 'Confirm secret code',
      type: 'secret',
      validation: [
        {
          rule: '$ == secretCode',
          message: 'Secret codes must match.'
        }
      ]
    },
    {
      name: 'firstName',
      label: 'First name',
      type: 'text',
      validation: [
        {
          rule: 'len($) > 5',
          message: 'First name must be longer than 5 characters.'
        }
      ],
    },
    {
      name: 'lastName',
      label: 'Last name',
      type: 'text',
      validation: [
        {
          rule: 'len($) > 5',
          message: 'Last name must be longer than 5 characters.'
        }
      ],
    }
  ]
}
```

The above form schema can be generated by some sophisticated UI, and then rendered on frontend, mobile devices etc with local validation + backend valdiation.

**Reply by prasunsrivastava** [reply]
This is fantastic. Thanks for showing me another use of the expression language. We are eagerly waiting for the android and iOS support.
